### PR TITLE
Added referral system

### PR DESCRIPTION
**Pull request by ztcrypto**

### CODE REVIEW

File: src/views/DPI/DPI.tsx [hunk 1]
```diff
@@ -37,6 +40,15 @@ const DpiProductPage = (props: { title: string }) => {
   const { components } = useDpiIndexComponents()
   const { dpiBalance } = useBalances()
 
+  const [_, setReferral] = useLocalStorage('referral', '')
+
+  const history = useHistory()
+  const params = new URLSearchParams(history.location.search)
+  const value = params.get('referral')
```
Comment (controtie) [inline src/views/DPI/DPI.tsx:47]: should this section be inside the useEffect below?

**Review by controtie — CHANGES_REQUESTED** [review]
Looks mostly clean and good. Minor comments. Gonna look into permissioning on the API key.

File: src/utils/referralApi.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import cred from 'serviceAccount.json'
+const { GoogleSpreadsheet } = require('google-spreadsheet')
+
+const addReferral = async (transactionId: string, referral: string) => {
```
[the remark this answers is not in the document]
Comment (ztcrypto) [reply src/utils/referralApi.ts:4]: Yes, current impl is like that. add an entry even it's empty.
What's the preference? add or not?

[the remark this answers is not in the document]
Comment (controtie) [reply src/utils/referralApi.ts:4]: Yup adding is fine.

**Comment by controtie** [comment]
Merged in #139

**Pull request closed without merging**